File: backend/app/schemas/stocks.py
from datetime import date

from pydantic import BaseModel, ConfigDict


class StockListItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    # 列表页仅暴露必要字段，避免过度拉取大字段。
    ts_code: str
    symbol: str
    name: str
    fullname: str | None
    exchange: str | None
    close: float | None
    pct_chg: float | None
    trade_date: date | None


class StockInstrumentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    ts_code: str
    symbol: str
    name: str
    area: str | None
    industry: str | None
    fullname: str | None
    enname: str | None
    cnspell: str | None
    market: str | None
    exchange: str | None
    curr_type: str | None
    list_status: str
    list_date: date | None
    delist_date: date | None
    is_hs: str | None
    act_name: str | None
    act_ent_type: str | None


class AdminStockPageResponse(BaseModel):
    items: list[StockInstrumentResponse]
    total: int
    page: int
    page_size: int


class StockDailySnapshotResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    ts_code: str
    trade_date: date
    open: float | None
    high: float | None
    low: float | None
    close: float | None
    pre_close: float | None
    change: float | None
    pct_chg: float | None
    vol: float | None
    amount: float | None
    turnover_rate: float | None
    volume_ratio: float | None
    pe: float | None
    pb: float | None
    total_mv: float | None
    circ_mv: float | None


class StockDetailResponse(BaseModel):
    # 详情页聚合证券基础信息与最近快照，快照可能为空。
    instrument: StockInstrumentResponse
    latest_snapshot: StockDailySnapshotResponse | None


class StockBasicSyncResponse(BaseModel):
    # 同步结果用于后台任务反馈，不影响前台业务数据。
    message: str
    total: int
    created: int
    updated: int
    list_statuses: list[str]


class StockTradeCalendarResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    exchange: str
    cal_date: date
    # is_open 为数据源原样字符串，保持口径一致。
    is_open: str
    pretrade_date: date | None


class StockAdjFactorResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    ts_code: str
    trade_date: date
    adj_factor: float

File: frontend/src/api/stocks.ts
import { requestJson } from './http'


export type StockListItem = {
  ts_code: string
  symbol: string
  name: string
  fullname: string | null
  exchange: string | null
  close: number | null
  pct_chg: number | null
  trade_date: string | null
}


export type StockInstrument = {
  ts_code: string
  symbol: string
  name: string
  area: string | null
  industry: string | null
  fullname: string | null
  enname: string | null
  cnspell: string | null
  market: string | null
  exchange: string | null
  curr_type: string | null
  list_status: string
  list_date: string | null
  delist_date: string | null
  is_hs: string | null
  act_name: string | null
  act_ent_type: string | null
}


export type StockDailySnapshot = {
  ts_code: string
  trade_date: string
  open: number | null
  high: number | null
  low: number | null
  close: number | null
  pre_close: number | null
  change: number | null
  pct_chg: number | null
  vol: number | null
  amount: number | null
  turnover_rate: number | null
  volume_ratio: number | null
  pe: number | null
  pb: number | null
  total_mv: number | null
  circ_mv: number | null
}


export type StockDetail = {
  instrument: StockInstrument
  latest_snapshot: StockDailySnapshot | null
}


const buildQueryString = (params: Record<string, string | number | undefined>) => {
  const query = new URLSearchParams()
  Object.entries(params).forEach(([key, value]) => {
    if (value === undefined || value === '') {
      return
    }
    query.set(key, String(value))
  })
  const queryString = query.toString()
  return queryString ? `?${queryString}` : ''
}


export const stocksApi = {
  async listStocks(keyword?: string, listStatus?: string, page = 1, pageSize = 20) {
    const query = buildQueryString({
      keyword,
      list_status: listStatus,
      page,
      page_size: pageSize,
    })
    return requestJson<StockListItem[]>(`/api/stocks${query}`)
  },
  async getStockDetail(tsCode: string) {
    return requestJson<StockDetail>(`/api/stocks/${encodeURIComponent(tsCode)}`)
  },
  async getStockDaily(tsCode: string, limit = 60) {
    const query = buildQueryString({ limit })
    return requestJson<StockDailySnapshot[]>(
      `/api/stocks/${encodeURIComponent(tsCode)}/daily${query}`,
    )
  },
}
